engine/src/main/java/org/entando/entando/web/common/interceptor/EntandoOauth2Interceptor.java

```diff
@@ -0,0 +1,117 @@
+import org.entando.entando.aps.system.exception.RestServerError;
+import org.entando.entando.aps.system.services.oauth2.IApiOAuth2TokenManager;
+import org.entando.entando.aps.system.services.oauth2.model.OAuth2Token;
+import org.entando.entando.web.common.annotation.RestAccessControl;
+import org.entando.entando.web.common.exceptions.EntandoAuthorizationException;
+import org.entando.entando.web.common.exceptions.EntandoTokenException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+/**
+ *
+ * @author paddeo
+ */
+public class EntandoOauth2Interceptor extends HandlerInterceptorAdapter {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private IApiOAuth2TokenManager oAuth2TokenManager;
+
+    @Autowired
+    private IAuthenticationProviderManager authenticationProviderManager;
+
+    @Autowired
+    private IAuthorizationManager authorizationManager;
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
+        if (handler instanceof HandlerMethod) {
+            HandlerMethod method = (HandlerMethod) handler;
+            if (method.getMethod().isAnnotationPresent(RequestMapping.class)) {
+                RestAccessControl rqm = method.getMethodAnnotation(RestAccessControl.class);
+                String permission = rqm.permission();
+                this.extractOAuthParameters(request, permission);
+            }
+        }
+        return true;
+    }
+
+    protected void extractOAuthParameters(HttpServletRequest request, String permission) {
+        try {
+            logger.info("Permission required: {}", permission);
+            OAuthAccessResourceRequest requestMessage = new OAuthAccessResourceRequest(request, ParameterStyle.HEADER);
+
+            String accessToken = requestMessage.getAccessToken();
+            if (StringUtils.isBlank(accessToken)) {
+                throw new EntandoTokenException("no access token found", request, null);
+            }
+
+            final OAuth2Token token = oAuth2TokenManager.getApiOAuth2Token(accessToken);
+            this.validateToken(request, accessToken, token);
+
+            String username = token.getClientId();
+            this.checkAuthorization(username, permission, request);
+
+        } catch (OAuthSystemException | ApsSystemException | OAuthProblemException ex) {
+            logger.error("System exception {}", ex.getMessage());
+            throw new RestServerError("error parsing OAuth parameters", ex);
+        }
+    }
+
+    protected void checkAuthorization(String username, String permission, HttpServletRequest request) throws ApsSystemException {
```

Comment:
It seems odd to me that this method doesn't have a return type. Are we relying on exceptions for all auth flow control? I get why but the method no-ops if the user isn't found which is the same as "authorized". So if i made up a fake user name i get to be superuser.

Reply:
you can not get to this point if you have not obtained a valid access token, ie if you are neither an authorized API consumer nor an authenticated Entando user